test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -408,18 +419,58 @@ void UcxContext::destroy_worker()
     ucp_worker_destroy(_worker);
 }
 
+ucs_status_t UcxContext::am_recv_callback(void *arg, const void *header,
+                                          size_t header_length,
+                                          void *data, size_t length,
+                                          const ucp_am_recv_param_t *param)
+{
+    UcxContext *self = reinterpret_cast<UcxContext*>(arg);
+
+    assert(param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP);
+
+    uint64_t id = reinterpret_cast<uint64_t>(param->reply_ep);
+    conn_map_t::iterator iter = self->_conns.find(id);
+    if (iter == self->_conns.end()) {
+        UCX_LOG << "could not find connection with ep " << param->reply_ep
+                << "(" << id << ")";
+        return UCS_OK;
+    }
+    UcxConnection *conn = iter->second;
+
+    conn->set_am_desc(data, param);
+
+    self->dispatch_am_message(conn, header, header_length);
+
+    return UCS_OK;
```

Comment:
can we safety return OK here if data and param pointer is assigned to conn above?
also need to reset `_am_desc` from `UcxConnection::recv_am_data` to avoid pointers to released data and structures.

Reply:
yes, because we do not have deferred receive. Messages are handled by `dispatch_am_message`
Ok, will reset

Reply:
pls add a comment,
but IMO it looks buggy prone, can we avoid storing temporary pointers to long living structure (conn) at all?

Reply:
trying to avoid exposing conn UCX info to IODEMO. This approach seems to be similar to handling of `_iomsg_recv_request`

Reply:
recv_request is always valid and handled separately

Reply:
recv_request is always changed, like the am_desc discussed here

Reply:
but it's handled and reposted from the same context, am_desc is valid only during callback
the diff is life time of the object, seems like you are storing values to connection which actually should be parameters of `dispatch_am_message` func